_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+
+# serve model (limiting number of requests)
+model$serve(num_requests = 3L)
+```
+
+Once the desired number of requests have been served, we can go to this R process, stop model sharing and shut down the
+cluster:
+
+```{r}
+# stop model sharing
+model$stop()
+
+# stop cluster
+cluster$stop()
+```
+
+Before that though, on to the client(s).
+
+#### Request predictions on private data
+
+In our example, we have one client. The client is a TFE worker, just like the agents that make up the cluster.
+
+We define the cluster here, client-side, as well; create the client; and connect the client to the model. This will set up a
+queueing server that takes care of *secret sharing* all input data before submitting them for prediction.
+
+Finally, we have the client asking for classification of the first three MNIST images.
+
+With the server running in some different R process, we can conveniently run this in RStudio:
+
+```{r}
+# client.R
+
+library(tensorflow)
+library(keras)
+
+sy <- reticulate::import(("syft"))
+hook <- sy$KerasHook(tf$keras)
+
+mnist <- dataset_mnist()
+mnist$train$x <- mnist$train$x/255
+mnist$test$x <- mnist$test$x/255
+
+dim(mnist$train$x) <- c(dim(mnist$train$x), 1)
+dim(mnist$test$x) <- c(dim(mnist$test$x), 1)
+
+batch_input_shape <- c(1, 28, 28, 1)
+batch_output_shape <- c(1, 10)
+
+# define the same TFE cluster
+AUTO <- TRUE
+julia <- sy$TFEWorker(host = 'localhost:4000', auto_managed = AUTO)
+greg <- sy$TFEWorker(host = 'localhost:4001', auto_managed = AUTO)
+me <- sy$TFEWorker(host = 'localhost:4002', auto_managed = AUTO)
+cluster <- sy$TFECluster(julia, greg, me)
+
+# create the client
+client <- sy$TFEWorker()
+
+# create a queueing server on the client that secret shares the data 
+# before submitting a prediction request
+client$connect_to_model(batch_input_shape, batch_output_shape, cluster)
+
+num_tests <- 3
+images <- mnist$test$x[1: num_tests, , , , drop = FALSE]
+expected_labels <- mnist$test$y[1: num_tests]
+
+for (i in 1:num_tests) {
+  res <- client$query_model(images[i, , , , drop = FALSE])
+  predicted_label <- which.max(res) - 1
+  cat("Actual: ", expected_labels[i], ", predicted: ", predicted_label)
+}
+  
+  
+```
+
+    Actual:  7 , predicted:  7 
+    Actual:  2 , predicted:  2 
+    Actual:  1 , predicted:  1 
+
+There we go. Both model and data did remain secret, yet we were able to classify our data.
+
+Let's wrap up.
+
+## Conclusion
+
+Our example use case has not been too ambitious -- we started with a trained model, thus leaving aside federated learning.
+Keeping the setup simple, we were able to focus on underlying principles: *Secret sharing* as a means of encryption, and
+setting up a Syft/TFE cluster of workers that together, provide the infrastructure for encrypting model weights as well as client data.
+
+In case you've read our previous post on [TensorFlow
+Federated](https://blogs.rstudio.com/tensorflow/posts/2019-12-20-differential-privacy/), that too a framework under
```

Comment:
```suggestion
Federated](https://blogs.rstudio.com/tensorflow/posts/2019-12-20-differential-privacy/), that too is a framework under
```

Reply:
hm, that's not what i intended ... that was meant as a parenthetical... maybe i should try em-dashes instead of commas

Reply:
changed to 
> – that, too, a framework under development –